src/Umbraco.Core/Services/UserGroupService.cs

```diff
@@ -0,0 +1,368 @@
+        }
+
+        if (await IsNewUserGroup(userGroup) is false)
+        {
+            return Attempt.FailWithStatus(UserGroupOperationStatus.AlreadyExists, userGroup);
+        }
+
+        return UserGroupHasUniqueAlias(userGroup) is false
+            ? Attempt.FailWithStatus(UserGroupOperationStatus.DuplicateAlias, userGroup)
+            : Attempt.SucceedWithStatus(UserGroupOperationStatus.Success, userGroup);
+    }
+
+    /// <inheritdoc />
+    public async Task<Attempt<IUserGroup, UserGroupOperationStatus>> UpdateAsync(
+        IUserGroup userGroup,
+        int performingUserId)
+    {
+        using ICoreScope scope = ScopeProvider.CreateCoreScope();
+
+        IUser? performingUser = _userService.GetUserById(performingUserId);
+        if (performingUser is null)
+        {
+            return Attempt.FailWithStatus(UserGroupOperationStatus.MissingUser, userGroup);
+        }
+
+        UserGroupOperationStatus validationStatus = await ValidateUserGroupUpdateAsync(userGroup);
+        if (validationStatus is not UserGroupOperationStatus.Success)
+        {
+            return Attempt.FailWithStatus(validationStatus, userGroup);
+        }
+
+        Attempt<UserGroupOperationStatus> authorizationAttempt = _userGroupAuthorizationService.AuthorizeUserGroupUpdate(performingUser, userGroup);
+        if (authorizationAttempt.Success is false)
+        {
+            return Attempt.FailWithStatus(authorizationAttempt.Result, userGroup);
+        }
+
+        EventMessages eventMessages = EventMessagesFactory.Get();
+        var savingNotification = new UserGroupSavingNotification(userGroup, eventMessages);
+        if (await scope.Notifications.PublishCancelableAsync(savingNotification))
+        {
+            scope.Complete();
+            return Attempt.FailWithStatus(UserGroupOperationStatus.CancelledByNotification, userGroup);
+        }
+
+        _userGroupRepository.Save(userGroup);
+        scope.Notifications.Publish(new UserGroupSavedNotification(userGroup, eventMessages).WithStateFrom(savingNotification));
+
+        scope.Complete();
+        return Attempt.SucceedWithStatus(UserGroupOperationStatus.Success, userGroup);
+    }
+
+    private async Task<UserGroupOperationStatus> ValidateUserGroupUpdateAsync(IUserGroup userGroup)
+    {
+        UserGroupOperationStatus commonValidationStatus = ValidateCommon(userGroup);
+        if (commonValidationStatus != UserGroupOperationStatus.Success)
+        {
+            return commonValidationStatus;
+        }
+
+        if (await IsNewUserGroup(userGroup))
+        {
+            return UserGroupOperationStatus.NotFound;
+        }
+
+        return UserGroupOperationStatus.Success;
+    }
+
+    /// <summary>
+    /// Validate common user group properties, that are shared between update, create, etc.
+    /// </summary>
+    private UserGroupOperationStatus ValidateCommon(IUserGroup userGroup)
+    {
+        if (userGroup.Name!.Length > MaxUserGroupNameLength)
```

Comment:
From a service perspective, I can pass in a usergroup with name=null. This will explode here?

Reply:
Example tests showing this:

```cs
using NUnit.Framework;
using Umbraco.Cms.Core;
using Umbraco.Cms.Core.Models.Membership;
using Umbraco.Cms.Core.Services;
using Umbraco.Cms.Core.Strings;
using Umbraco.Cms.Tests.Common.Testing;
using Umbraco.Cms.Tests.Integration.Testing;

namespace Umbraco.Cms.Tests.Integration.Umbraco.Core.Services;


[TestFixture]
[UmbracoTest(Database = UmbracoTestOptions.Database.NewSchemaPerTest)]
public class UserGroupServiceTests : UmbracoIntegrationTest
{
    private IUserGroupService UserGroupService => GetRequiredService<IUserGroupService>();
    private IShortStringHelper ShortStringHelper => GetRequiredService<IShortStringHelper>();

    [Test]
    public async Task Cannot_create_user_group_with_name_equals_null()
    {

        var userGroup = new UserGroup(ShortStringHelper)
        {
            Name = null
        };
        
        var result = await UserGroupService.CreateAsync(userGroup, Constants.Security.SuperUserId);
        
        Assert.IsFalse(result.Success);
    }
}
```

Reply:
Very good point, I've fixed this now, I've added the check as a `string.IsNullOrEmpty`, since otherwise you could create a nameless group 🙈